File: liyaqa-member-app/shared/src/commonMain/kotlin/com/liyaqa/member/domain/model/Wearable.kt
package com.liyaqa.member.domain.model

import kotlinx.serialization.Serializable

/**
 * Wearable platform names matching backend
 */
enum class WearablePlatformName {
    FITBIT,
    GARMIN,
    GOOGLE_FIT,
    APPLE_HEALTH,
    WHOOP,
    OURA;

    fun getDisplayName(isArabic: Boolean): String = when (this) {
        FITBIT -> if (isArabic) "فيتبيت" else "Fitbit"
        GARMIN -> if (isArabic) "جارمين كونكت" else "Garmin Connect"
        GOOGLE_FIT -> if (isArabic) "جوجل فيت" else "Google Fit"
        APPLE_HEALTH -> if (isArabic) "صحة أبل" else "Apple Health"
        WHOOP -> if (isArabic) "ووب" else "WHOOP"
        OURA -> if (isArabic) "خاتم أورا" else "Oura Ring"
    }
}

enum class WearableAuthType {
    OAUTH2,
    DEVICE_SDK
}

enum class SyncSource {
    API,
    WEBHOOK,
    SDK
}

enum class SyncStatus {
    SUCCESS,
    FAILED,
    PARTIAL;

    fun getLabel(isArabic: Boolean): String = when (this) {
        SUCCESS -> if (isArabic) "ناجح" else "Success"
        FAILED -> if (isArabic) "فاشل" else "Failed"
        PARTIAL -> if (isArabic) "جزئي" else "Partial"
    }
}

enum class SyncJobType {
    FULL_SYNC,
    INCREMENTAL,
    ACTIVITIES,
    WORKOUTS
}

enum class SyncJobStatus {
    PENDING,
    RUNNING,
    COMPLETED,
    FAILED;

    fun getLabel(isArabic: Boolean): String = when (this) {
        PENDING -> if (isArabic) "في الانتظار" else "Pending"
        RUNNING -> if (isArabic) "قيد التشغيل" else "Running"
        COMPLETED -> if (isArabic) "مكتمل" else "Completed"
        FAILED -> if (isArabic) "فاشل" else "Failed"
    }
}

enum class WearableActivityType {
    RUNNING,
    WALKING,
    CYCLING,
    SWIMMING,
    GYM_WORKOUT,
    HIIT,
    YOGA,
    PILATES,
    STRENGTH_TRAINING,
    CARDIO,
    ELLIPTICAL,
    ROWING,
    STAIR_CLIMBING,
    TREADMILL,
    CROSS_TRAINING,
    HIKING,
    DANCE,
    MARTIAL_ARTS,
    STRETCHING,
    MEDITATION,
    SLEEP,
    OTHER;

    fun getLabel(isArabic: Boolean): String = when (this) {
        RUNNING -> if (isArabic) "الجري" else "Running"
        WALKING -> if (isArabic) "المشي" else "Walking"
        CYCLING -> if (isArabic) "ركوب الدراجة" else "Cycling"
        SWIMMING -> if (isArabic) "السباحة" else "Swimming"
        GYM_WORKOUT -> if (isArabic) "تمرين صالة" else "Gym Workout"
        HIIT -> if (isArabic) "تدريب متقطع" else "HIIT"
        YOGA -> if (isArabic) "يوغا" else "Yoga"
        PILATES -> if (isArabic) "بيلاتس" else "Pilates"
        STRENGTH_TRAINING -> if (isArabic) "تدريب القوة" else "Strength Training"
        CARDIO -> if (isArabic) "كارديو" else "Cardio"
        ELLIPTICAL -> if (isArabic) "إليبتيكال" else "Elliptical"
        ROWING -> if (isArabic) "التجديف" else "Rowing"
        STAIR_CLIMBING -> if (isArabic) "صعود الدرج" else "Stair Climbing"
        TREADMILL -> if (isArabic) "جهاز المشي" else "Treadmill"
        CROSS_TRAINING -> if (isArabic) "تدريب شامل" else "Cross Training"
        HIKING -> if (isArabic) "المشي لمسافات" else "Hiking"
        DANCE -> if (isArabic) "الرقص" else "Dance"
        MARTIAL_ARTS -> if (isArabic) "فنون قتالية" else "Martial Arts"
        STRETCHING -> if (isArabic) "التمدد" else "Stretching"
        MEDITATION -> if (isArabic) "التأمل" else "Meditation"
        SLEEP -> if (isArabic) "النوم" else "Sleep"
        OTHER -> if (isArabic) "أخرى" else "Other"
    }

    companion object {
        fun fromString(value: String): WearableActivityType =
            entries.find { it.name == value } ?: OTHER
    }
}

/**
 * Wearable platform configuration
 */
@Serializable
data class WearablePlatform(
    val id: String,
    val name: String,
    val displayName: String,
    val authType: String,
    val logoUrl: String?,
    val isActive: Boolean,
    val supportsOAuth: Boolean
) {
    val platformName: WearablePlatformName?
        get() = try {
            WearablePlatformName.valueOf(name)
        } catch (e: Exception) {
            null
        }

    val authTypeEnum: WearableAuthType
        get() = try {
            WearableAuthType.valueOf(authType)
        } catch (e: Exception) {
            WearableAuthType.OAUTH2
        }

    val isDeviceSdk: Boolean
        get() = authTypeEnum == WearableAuthType.DEVICE_SDK
}

/**
 * Member's wearable connection
 */
@Serializable
data class WearableConnection(
    val id: String,
    val memberId: String,
    val platformId: String,
    val platformName: String?,
    val platformDisplayName: String?,
    val platformLogoUrl: String?,
    val externalUserId: String?,
    val externalUsername: String?,
    val hasOAuthTokens: Boolean,
    val isTokenExpired: Boolean,
    val syncEnabled: Boolean,
    val lastSyncAt: String?,
    val lastSyncStatus: String?,
    val createdAt: String,
    val updatedAt: String
) {
    val platformEnum: WearablePlatformName?
        get() = platformName?.let {
            try {
                WearablePlatformName.valueOf(it)
            } catch (e: Exception) {
                null
            }
        }

    val syncStatusEnum: SyncStatus?
        get() = lastSyncStatus?.let {
            try {
                SyncStatus.valueOf(it)
            } catch (e: Exception) {
                null
            }
        }
}

/**
 * Daily activity summary from wearable
 */
@Serializable
data class WearableDailyActivity(
    val id: String,
    val memberId: String,
    val connectionId: String,
    val activityDate: String,
    val steps: Int?,
    val distanceMeters: Int?,
    val distanceKm: Double?,
    val floorsClimbed: Int?,
    val caloriesTotal: Int?,
    val caloriesActive: Int?,
    val activeMinutes: Int?,
    val activeHours: Double?,
    val sedentaryMinutes: Int?,
    val sleepMinutes: Int?,
    val sleepHours: Double?,
    val sleepQualityScore: Int?,
    val restingHeartRate: Int?,
    val hrvAverage: Double?,
    val stressScore: Int?,
    val recoveryScore: Int?,
    val syncSource: String,
    val createdAt: String
)

/**
 * Individual workout from wearable
 */
@Serializable
data class WearableWorkout(
    val id: String,
    val memberId: String,
    val connectionId: String,
    val externalWorkoutId: String?,
    val activityType: String,
    val activityName: String?,
    val startedAt: String,
    val endedAt: String?,
    val durationSeconds: Int?,
    val durationMinutes: Double?,
    val distanceMeters: Int?,
    val distanceKm: Double?,
    val caloriesBurned: Int?,
    val avgHeartRate: Int?,
    val maxHeartRate: Int?,
    val avgPaceSecondsPerKm: Int?,
    val avgPaceMinutesPerKm: Double?,
    val elevationGainMeters: Int?,
    val steps: Int?,
    val syncSource: String,
    val createdAt: String
) {
    val activityTypeEnum: WearableActivityType
        get() = WearableActivityType.fromString(activityType)
}

/**
 * Workout statistics aggregation
 */
@Serializable
data class WearableWorkoutStats(
    val totalWorkouts: Int,
    val totalDurationSeconds: Long,
    val totalDurationMinutes: Double,
    val totalDurationHours: Double,
    val totalCalories: Long
)

/**
 * Activity statistics aggregation
 */
@Serializable
data class WearableActivityStats(
    val daysTracked: Int,
    val totalSteps: Long,
    val averageStepsPerDay: Double,
    val totalCalories: Long,
    val totalActiveMinutes: Long,
    val totalActiveHours: Double,
    val averageSleepMinutes: Double,
    val averageSleepHours: Double,
    val averageRestingHeartRate: Double?
)

/**
 * Sync job tracking
 */
@Serializable
data class WearableSyncJob(
    val id: String,
    val connectionId: String,
    val jobType: String,
    val status: String,
    val startedAt: String?,
    val completedAt: String?,
    val durationSeconds: Int?,
    val recordsProcessed: Int,
    val recordsCreated: Int,
    val recordsUpdated: Int,
    val recordsFailed: Int,
    val errorMessage: String?,
    val createdAt: String
) {
    val jobTypeEnum: SyncJobType
        get() = try {
            SyncJobType.valueOf(jobType)
        } catch (e: Exception) {
            SyncJobType.FULL_SYNC
        }

    val statusEnum: SyncJobStatus
        get() = try {
            SyncJobStatus.valueOf(status)
        } catch (e: Exception) {
            SyncJobStatus.PENDING
        }
}

File: frontend/shared/src/types/wearable.ts
import type { UUID } from "./api";

// Enums
export type WearablePlatformName =
  | "FITBIT"
  | "GARMIN"
  | "GOOGLE_FIT"
  | "APPLE_HEALTH"
  | "WHOOP"
  | "OURA";

export type WearableAuthType = "OAUTH2" | "DEVICE_SDK";
export type SyncSource = "API" | "WEBHOOK" | "SDK";
export type SyncJobType = "FULL_SYNC" | "INCREMENTAL" | "ACTIVITIES" | "WORKOUTS";
export type SyncJobStatus = "PENDING" | "RUNNING" | "COMPLETED" | "FAILED";
export type SyncStatus = "SUCCESS" | "FAILED" | "PARTIAL";

export type WearableActivityType =
  | "RUNNING"
  | "WALKING"
  | "CYCLING"
  | "SWIMMING"
  | "GYM_WORKOUT"
  | "HIIT"
  | "YOGA"
  | "PILATES"
  | "STRENGTH_TRAINING"
  | "CARDIO"
  | "ELLIPTICAL"
  | "ROWING"
  | "STAIR_CLIMBING"
  | "TREADMILL"
  | "CROSS_TRAINING"
  | "HIKING"
  | "DANCE"
  | "MARTIAL_ARTS"
  | "STRETCHING"
  | "MEDITATION"
  | "SLEEP"
  | "OTHER";

// ========== Request Types ==========

export interface CreateConnectionRequest {
  memberId: UUID;
  platformId: UUID;
  externalUserId?: string;
  externalUsername?: string;
}

export interface UpdateConnectionRequest {
  externalUserId?: string;
  externalUsername?: string;
  syncEnabled?: boolean;
}

export interface UpdateConnectionTokensRequest {
  accessToken: string;
  refreshToken?: string;
  expiresAt?: string;
}

export interface CreateDailyActivityRequest {
  memberId: UUID;
  connectionId: UUID;
  activityDate: string;
  steps?: number;
  distanceMeters?: number;
  floorsClimbed?: number;
  caloriesTotal?: number;
  caloriesActive?: number;
  activeMinutes?: number;
  sedentaryMinutes?: number;
  sleepMinutes?: number;
  sleepQualityScore?: number;
  restingHeartRate?: number;
  hrvAverage?: number;
  stressScore?: number;
  recoveryScore?: number;
  rawData?: Record<string, unknown>;
}

export interface CreateWearableWorkoutRequest {
  memberId: UUID;
  connectionId: UUID;
  externalWorkoutId?: string;
  activityType: string;
  activityName?: string;
  startedAt: string;
  endedAt?: string;
  durationSeconds?: number;
  distanceMeters?: number;
  caloriesBurned?: number;
  avgHeartRate?: number;
  maxHeartRate?: number;
  avgPaceSecondsPerKm?: number;
  elevationGainMeters?: number;
  steps?: number;
  rawData?: Record<string, unknown>;
}

export interface StartSyncRequest {
  jobType?: SyncJobType;
}

// ========== Response Types ==========

export interface WearablePlatform {
  id: UUID;
  name: WearablePlatformName;
  displayName: string;
  apiBaseUrl: string | null;
  oauthAuthUrl: string | null;
  authType: WearableAuthType;
  logoUrl: string | null;
  isActive: boolean;
  supportsOAuth: boolean;
}

export interface MemberWearableConnection {
  id: UUID;
  memberId: UUID;
  platformId: UUID;
  platformName: WearablePlatformName | null;
  platformDisplayName: string | null;
  platformLogoUrl: string | null;
  externalUserId: string | null;
  externalUsername: string | null;
  hasOAuthTokens: boolean;
  isTokenExpired: boolean;
  syncEnabled: boolean;
  lastSyncAt: string | null;
  lastSyncStatus: SyncStatus | null;
  createdAt: string;
  updatedAt: string;
}

export interface WearableDailyActivity {
  id: UUID;
  memberId: UUID;
  connectionId: UUID;
  activityDate: string;
  steps: number | null;
  distanceMeters: number | null;
  distanceKm: number | null;
  floorsClimbed: number | null;
  caloriesTotal: number | null;
  caloriesActive: number | null;
  activeMinutes: number | null;
  activeHours: number | null;
  sedentaryMinutes: number | null;
  sleepMinutes: number | null;
  sleepHours: number | null;
  sleepQualityScore: number | null;
  restingHeartRate: number | null;
  hrvAverage: number | null;
  stressScore: number | null;
  recoveryScore: number | null;
  syncSource: SyncSource;
  createdAt: string;
}

export interface WearableWorkout {
  id: UUID;
  memberId: UUID;
  connectionId: UUID;
  externalWorkoutId: string | null;
  activityType: string;
  activityName: string | null;
  startedAt: string;
  endedAt: string | null;
  durationSeconds: number | null;
  durationMinutes: number | null;
  distanceMeters: number | null;
  distanceKm: number | null;
  caloriesBurned: number | null;
  avgHeartRate: number | null;
  maxHeartRate: number | null;
  avgPaceSecondsPerKm: number | null;
  avgPaceMinutesPerKm: number | null;
  elevationGainMeters: number | null;
  steps: number | null;
  syncSource: SyncSource;
  createdAt: string;
}

export interface WearableWorkoutStats {
  totalWorkouts: number;
  totalDurationSeconds: number;
  totalDurationMinutes: number;
  totalDurationHours: number;
  totalCalories: number;
}

export interface WearableActivityStats {
  daysTracked: number;
  totalSteps: number;
  averageStepsPerDay: number;
  totalCalories: number;
  totalActiveMinutes: number;
  totalActiveHours: number;
  averageSleepMinutes: number;
  averageSleepHours: number;
  averageRestingHeartRate: number | null;
}

export interface WearableSyncJob {
  id: UUID;
  connectionId: UUID;
  jobType: SyncJobType;
  status: SyncJobStatus;
  startedAt: string | null;
  completedAt: string | null;
  durationSeconds: number | null;
  recordsProcessed: number;
  recordsCreated: number;
  recordsUpdated: number;
  recordsFailed: number;
  errorMessage: string | null;
  createdAt: string;
}

// ========== Labels ==========

export const WEARABLE_PLATFORM_LABELS: Record<WearablePlatformName, string> = {
  FITBIT: "Fitbit",
  GARMIN: "Garmin Connect",
  GOOGLE_FIT: "Google Fit",
  APPLE_HEALTH: "Apple Health",
  WHOOP: "WHOOP",
  OURA: "Oura Ring",
};

export const WEARABLE_PLATFORM_LABELS_AR: Record<WearablePlatformName, string> = {
  FITBIT: "فيتبيت",
  GARMIN: "جارمين كونكت",
  GOOGLE_FIT: "جوجل فيت",
  APPLE_HEALTH: "صحة أبل",
  WHOOP: "ووب",
  OURA: "خاتم أورا",
};

export const SYNC_STATUS_LABELS: Record<SyncStatus, string> = {
  SUCCESS: "Success",
  FAILED: "Failed",
  PARTIAL: "Partial",
};

export const SYNC_STATUS_LABELS_AR: Record<SyncStatus, string> = {
  SUCCESS: "ناجح",
  FAILED: "فاشل",
  PARTIAL: "جزئي",
};

export const SYNC_JOB_STATUS_LABELS: Record<SyncJobStatus, string> = {
  PENDING: "Pending",
  RUNNING: "Running",
  COMPLETED: "Completed",
  FAILED: "Failed",
};

export const SYNC_JOB_STATUS_LABELS_AR: Record<SyncJobStatus, string> = {
  PENDING: "في الانتظار",
  RUNNING: "قيد التشغيل",
  COMPLETED: "مكتمل",
  FAILED: "فاشل",
};

export const ACTIVITY_TYPE_LABELS: Record<WearableActivityType, string> = {
  RUNNING: "Running",
  WALKING: "Walking",
  CYCLING: "Cycling",
  SWIMMING: "Swimming",
  GYM_WORKOUT: "Gym Workout",
  HIIT: "HIIT",
  YOGA: "Yoga",
  PILATES: "Pilates",
  STRENGTH_TRAINING: "Strength Training",
  CARDIO: "Cardio",
  ELLIPTICAL: "Elliptical",
  ROWING: "Rowing",
  STAIR_CLIMBING: "Stair Climbing",
  TREADMILL: "Treadmill",
  CROSS_TRAINING: "Cross Training",
  HIKING: "Hiking",
  DANCE: "Dance",
  MARTIAL_ARTS: "Martial Arts",
  STRETCHING: "Stretching",
  MEDITATION: "Meditation",
  SLEEP: "Sleep",
  OTHER: "Other",
};

export const ACTIVITY_TYPE_LABELS_AR: Record<WearableActivityType, string> = {
  RUNNING: "الجري",
  WALKING: "المشي",
  CYCLING: "ركوب الدراجة",
  SWIMMING: "السباحة",
  GYM_WORKOUT: "تمرين صالة",
  HIIT: "تدريب متقطع",
  YOGA: "يوغا",
  PILATES: "بيلاتس",
  STRENGTH_TRAINING: "تدريب القوة",
  CARDIO: "كارديو",
  ELLIPTICAL: "إليبتيكال",
  ROWING: "التجديف",
  STAIR_CLIMBING: "صعود الدرج",
  TREADMILL: "جهاز المشي",
  CROSS_TRAINING: "تدريب شامل",
  HIKING: "المشي لمسافات",
  DANCE: "الرقص",
  MARTIAL_ARTS: "فنون قتالية",
  STRETCHING: "التمدد",
  MEDITATION: "التأمل",
  SLEEP: "النوم",
  OTHER: "أخرى",
};

// Helper to get activity type label (handles dynamic types from API)
export function getActivityTypeLabel(activityType: string, locale: "en" | "ar" = "en"): string {
  const labels = locale === "ar" ? ACTIVITY_TYPE_LABELS_AR : ACTIVITY_TYPE_LABELS;
  return labels[activityType as WearableActivityType] || activityType;
}
